src/main/java/blusunrize/immersiveengineering/api/tool/ConveyorHandler.java

```diff
@@ -363,57 +398,101 @@ else if(getConveyorDirection()==ConveyorDirection.DOWN)
 		{
 			ConveyorDirection conveyorDirection = getConveyorDirection();
 			BlockPos pos = getTile().getPos();
 
+			final Direction facing = getFacing();
 			double vBase = 1.15;
-			double vX = 0.1*vBase*getFacing().getXOffset();
+			double vX = 0.1*vBase*facing.getXOffset();
 			double vY = entity.getMotion().y;
-			double vZ = 0.1*vBase*getFacing().getZOffset();
+			double vZ = 0.1*vBase*facing.getZOffset();
 
-			if(conveyorDirection==ConveyorDirection.UP)
-				vY = 0.17D*vBase;
-			else if(conveyorDirection==ConveyorDirection.DOWN)
-				vY = -0.07*vBase;
-
-			if(conveyorDirection!=ConveyorDirection.HORIZONTAL)
-				entity.onGround = false;
-
-			if(getFacing()==Direction.WEST||getFacing()==Direction.EAST)
+			if(facing==Direction.WEST||facing==Direction.EAST)
 			{
 				if(entity.posZ > pos.getZ()+0.55D)
 					vZ = -0.1D*vBase;
 				else if(entity.posZ < pos.getZ()+0.45D)
 					vZ = 0.1D*vBase;
 			}
-			else if(getFacing()==Direction.NORTH||getFacing()==Direction.SOUTH)
+			else if(facing==Direction.NORTH||facing==Direction.SOUTH)
 			{
 				if(entity.posX > pos.getX()+0.55D)
 					vX = -0.1D*vBase;
 				else if(entity.posX < pos.getX()+0.45D)
 					vX = 0.1D*vBase;
 			}
 
+			if(conveyorDirection!=ConveyorDirection.HORIZONTAL)
+			{
+				// Attempt to fix entity to the highest point under it
+				final Vec3d centerRelative = entity.getPositionVec()
+						.subtract(new Vec3d(pos))
+						.subtract(0.5+vX, 0.5, 0.5+vZ);
+				final double conveyorHeight = 2/16.;
+				final double centerOffsetInDirection = centerRelative.dotProduct(new Vec3d(facing.getDirectionVec()));
+				final double radius = entity.getSize(entity.getPose()).width/2;
+				final double maxEntityPos = centerOffsetInDirection+radius;
+				double maxCenterHeightUnderEntity = maxEntityPos+conveyorHeight;
+				if(conveyorDirection==ConveyorDirection.DOWN)
+					maxCenterHeightUnderEntity = -maxCenterHeightUnderEntity;
+				if(conveyorDirection==ConveyorDirection.UP)
+				{
+					if(maxCenterHeightUnderEntity > centerRelative.y||!outputBlocked)
+						vY = 0.17D*vBase;
+				}
+				else
+					vY = Math.signum(maxCenterHeightUnderEntity-centerRelative.y)*0.07*vBase;
+				entity.onGround = false;
+			}
+
 			return new Vec3d(vX, vY, vZ);
 		}
 
 		default void onEntityCollision(Entity entity)
 		{
-			if(!isActive())
+			if(!isActive()||entity==null)
 				return;
+			boolean hasBeenHandled = !markEntityAsHandled(entity);
 			BlockPos pos = getTile().getPos();
 			ConveyorDirection conveyorDirection = getConveyorDirection();
 			float heightLimit = conveyorDirection==ConveyorDirection.HORIZONTAL?.25f: 1f;
-			if(entity!=null&&entity.isAlive()&&!(entity instanceof PlayerEntity&&entity.isSneaking())&&entity.posY-pos.getY() >= 0&&entity.posY-pos.getY() < heightLimit)
+			final boolean outputBlocked = isOutputBlocked();
+			if(entity.isAlive()&&!(entity instanceof PlayerEntity&&entity.isSneaking())&&entity.posY-pos.getY() >= 0&&entity.posY-pos.getY() < heightLimit)
```

Comment:
This if condition should be broken up for readability. Checks for "isAlive" and the player check can be changed to return earlier.

Reply:
I mean, it's already an improvement over what was there before. The only reason I moved the null-check is that I access the entity before that, so it can't ever be null at that point. But yes, I'll change it.